### PR TITLE
router: do not reuse retry plugin instances

### PR DESCRIPTION
**Pull request by snowp**
Currently each route with a configured retry policy instantiates a
single instance of the desired retry plugins. This makes the plugins
somewhat unusable because it's unable to differentiate between retries
happening for different requests.

This changes them to be instantiated once per retry state.

Refactors the retry plugin factory interface to be easier to work with.

Signed-off-by: Snow Pettersen <snowp@squareup.com>

*Risk Level*: Medium
*Testing*: Added unit test to ensure we're not returning the same pointer
*Docs Changes*: n/a
*Release Notes*: n/a

### CODE REVIEW

**Comment by snowp** [comment]
For context this was causing our usage of the `other_priorities` plugin to sometimes get multiple `onHostAttempted` in between `onDeterminePriorityLoad`, so it would sometimes skip the update and return an empty `PriorityLoad` which would trigger a NOT_REACHED assert

File: source/common/router/retry_state_impl.h [hunk 1]
```diff
@@ -41,8 +41,11 @@ class RetryStateImpl : public RetryState {
                           DoRetryCallback callback) override;
 
   void onHostAttempted(Upstream::HostDescriptionConstSharedPtr host) override {
-    std::for_each(retry_host_predicates_.begin(), retry_host_predicates_.end(),
-                  [&host](auto predicate) { predicate->onHostAttempted(host); });
+    std::for_each(
+        retry_host_predicates_.begin(), retry_host_predicates_.end(), [&host](auto predicate) {
+          std::cerr << "ptr: " << reinterpret_cast<uintptr_t>(predicate.get()) << std::endl;
```
Comment (alyssawilk) [inline source/common/router/retry_state_impl.h]: left over debug cerr?

File: source/common/router/config_impl.h [hunk 2]
```diff
@@ -185,42 +185,29 @@ typedef std::shared_ptr<VirtualHostImpl> VirtualHostSharedPtr;
...
-  }
-
 private:
   std::chrono::milliseconds per_try_timeout_{0};
   uint32_t num_retries_{};
   uint32_t retry_on_{};
-  std::vector<Upstream::RetryHostPredicateSharedPtr> retry_host_predicates_;
+  std::vector<std::pair<std::string, ProtobufTypes::MessagePtr>> retry_host_predicate_configs_;
   Upstream::RetryPrioritySharedPtr retry_priority_;
+  std::pair<std::string, ProtobufTypes::MessagePtr> retry_priority_config_;
```
Comment (alyssawilk) [inline source/common/router/config_impl.h]: Mind adding comments on these two new variables?

File: source/common/router/retry_state_impl.h (hunk 1 above)
Comment (snowp) [reply source/common/router/retry_state_impl.h]: totally, thanks for catching that

**Comment by snowp** [comment]
I'll do a quick scan over the comments but I believe so, considering I thought this was the case when I wrote the code :)

**Comment by enzoferrari** [comment]
Tell me more..

**Pull request merged**